File: mobile/lib/shared/widgets/connection_status_pill.dart
import 'package:flutter/material.dart';

import '../../core/theme/joi_colors.dart';
import '../../services/websocket/ws_service.dart';
import 'glass_card.dart';

class ConnectionStatusPill extends StatelessWidget {
  final WsConnectionState state;

  const ConnectionStatusPill({super.key, required this.state});

  @override
  Widget build(BuildContext context) {
    final (color, label) = switch (state) {
      WsConnectionState.connected => (JoiColors.success, 'Connected'),
      WsConnectionState.connecting => (JoiColors.warning, 'Connecting...'),
      WsConnectionState.reconnecting => (JoiColors.warning, 'Reconnecting...'),
      WsConnectionState.disconnected => (JoiColors.textTertiary, 'Offline'),
    };

    return GlassCard(
      blurSigma: 12,
      backgroundOpacity: 0.04,
      borderRadius: 999,
      padding: const EdgeInsets.symmetric(horizontal: 12, vertical: 6),
      child: Row(
        mainAxisSize: MainAxisSize.min,
        children: [
          _PulsingDot(
            color: color,
            pulse: state == WsConnectionState.connecting ||
                state == WsConnectionState.reconnecting,
          ),
          const SizedBox(width: 8),
          Text(
            label,
            style: TextStyle(
              color: color,
              fontSize: 12,
              fontWeight: FontWeight.w500,
            ),
          ),
        ],
      ),
    );
  }
}

class _PulsingDot extends StatefulWidget {
  final Color color;
  final bool pulse;

  const _PulsingDot({required this.color, this.pulse = false});

  @override
  State<_PulsingDot> createState() => _PulsingDotState();
}

class _PulsingDotState extends State<_PulsingDot>
    with SingleTickerProviderStateMixin {
  late AnimationController _controller;

  @override
  void initState() {
    super.initState();
    _controller = AnimationController(
      vsync: this,
      duration: const Duration(milliseconds: 1000),
    );
    if (widget.pulse) _controller.repeat(reverse: true);
  }

  @override
  void didUpdateWidget(_PulsingDot old) {
    super.didUpdateWidget(old);
    if (widget.pulse && !_controller.isAnimating) {
      _controller.repeat(reverse: true);
    } else if (!widget.pulse) {
      _controller.stop();
      _controller.value = 1.0;
    }
  }

  @override
  Widget build(BuildContext context) {
    return AnimatedBuilder(
      animation: _controller,
      builder: (context, _) => Container(
        width: 8,
        height: 8,
        decoration: BoxDecoration(
          shape: BoxShape.circle,
          color: widget.color.withValues(alpha: 0.5 + _controller.value * 0.5),
          boxShadow: [
            BoxShadow(
              color: widget.color.withValues(alpha: 0.3),
              blurRadius: 6,
            ),
          ],
        ),
      ),
    );
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }
}

File: app/JOI/Shared/ConnectionStatusPill.swift
import SwiftUI

struct ConnectionStatusPill: View {
    let state: WebSocketClient.ConnectionState
    @State private var isPulsing = false

    var body: some View {
        HStack(spacing: 6) {
            Circle()
                .fill(dotColor)
                .frame(width: 8, height: 8)
                .scaleEffect(isPulsing ? 1.3 : 1.0)
                .animation(
                    state == .connecting || state == .reconnecting
                        ? .easeInOut(duration: 0.8).repeatForever(autoreverses: true)
                        : .default,
                    value: isPulsing)

            Text(label)
                .font(JOITypography.labelSmall)
                .foregroundStyle(JOIColors.textSecondary)
        }
        .padding(.horizontal, 10)
        .padding(.vertical, 4)
        .background(JOIColors.surfaceVariant)
        .clipShape(Capsule())
        .onAppear {
            isPulsing = state == .connecting || state == .reconnecting
        }
        .onChange(of: state) {
            isPulsing = state == .connecting || state == .reconnecting
        }
    }

    private var dotColor: Color {
        switch state {
        case .connected: JOIColors.success
        case .connecting, .reconnecting: JOIColors.warning
        case .disconnected: JOIColors.error
        }
    }

    private var label: String {
        switch state {
        case .connected: "Voice Live"
        case .connecting: "Linking Voice"
        case .reconnecting: "Re-linking"
        case .disconnected: "Voice Offline"
        }
    }
}
